openedx_lti_tool_plugin/deep_linking/forms.py

```diff
@@ -1,131 +1,208 @@
-        lti_tool: LtiTool,
+        launch_data: dict,
         **kwargs: dict,
     ):
         """Class __init__ method.
 
-        Initialize class instance attributes and add field choices
-        to the `content_items` field.
+        Initialize class instance attributes and set the choices
+        of the content_items field.
 
         Args:
             *args: Variable length argument list.
             request: HttpRequest object.
-            lti_tool: LtiTool model instance.
+            launch_data: Launch message data.
             **kwargs: Arbitrary keyword arguments.
 
         """
         super().__init__(*args, **kwargs)
         self.request = request
-        self.lti_tool = lti_tool
+        self.launch_data = launch_data
         self.fields['content_items'].choices = self.get_content_items_choices()
 
-    def get_content_items_choices(self) -> Set[Optional[Tuple[str, str]]]:
-        """Get `content_items` field choices.
+    def get_content_items_choices(self) -> List[Optional[Tuple[str, str]]]:
+        """Get content_items field choices.
+
+        This method will obtain the choices from the list of content items
+        dictionaries provided by the get_content_items function.
+
+        If a content items provider function is setup, and this function
+        returns a list of content items, these content items will be used
+        instead, if the provider doesn't return any content items, it
+        will fallback to using the get_content_items function.
 
         Returns:
-            Set of tuples with choices for the `content_items` field or an empty set.
+            A list of tuples with content_items field choices or empty list.
 
         """
-        return {
-            self.get_content_items_choice(course)
-            for course in self.get_course_contexts()
-        }
+        return [
+            (json.dumps(content_item), content_item.get('title', ''))
+            for content_item in (
+                self.get_content_items_from_provider()
+                or self.get_content_items()
+            )
+        ]
 
-    def get_content_items_choice(self, course) -> Tuple[str, str]:
-        """Get `content_items` field choice.
+    def get_content_items_from_provider(self) -> List[Optional[dict]]:
+        """Get content items from a provider function.
 
-        Args:
-            course (CourseContext): Course object.
+        This method will try to obtain content items from a provider function.
+        To setup a provider function the OLTITP_DL_CONTENT_ITEMS_PROVIDER setting
+        must be set to a string with the full path to the function:
+
+        Example:
+            OLTITP_DL_CONTENT_ITEMS_PROVIDER = 'example.module.path.provider_function'
+
+        This method will then try to call the provider function, the call will include
+        the HTTPRequest object and deep linking launch data dictionary received from
+        the deep linking request has arguments.
```

Comment:
Why is it necessary that the function receives the request and the launch data? Wouldn't it be enough with specifying the content items in the settings as a dictionary?

Reply:
@alexjmpb We send the request and launch_data to a function and get the content items from a function instead of defining them on a setting because we need to dynamically obtain these content items, our custom use case for example is to obtain the classes of an institution